src/cli/commands/upgrade-interactive.js

```diff
@@ -137,9 +137,11 @@ export async function run(config: Config, reporter: Reporter, flags: Object, arg
       flags.dev = hint === 'dev';
       flags.peer = hint === 'peer';
       flags.optional = hint === 'optional';
-
       const deps = answers.filter(isHint(hint)).map(getPattern);
       if (deps.length) {
+        for (const pattern of deps) {
+          lockfile.removePattern(pattern);
+        }
```

Comment:
upgrade-interactive seemed to have been previously broken, regardless of workspaces or not. When `--latest` flag wasn't being passed, it needed https://github.com/yarnpkg/yarn/blob/master/src/cli/commands/upgrade.js#L84-L86, like upgrade, otherwise nothing would get re-installed.